sql/src/main/java/io/crate/metadata/Routing.java

```diff
@@ -128,6 +129,64 @@ public boolean containsShards(String nodeId) {
         return false;
     }
 
+    /**
+     * Update the locations of the current routing by merging them with the locations
+     * that have been computed in an existing routing for the same table.
+     * <p>
+     * if a shard has been already routed in {@paramref existingLocations}
+     * then: use the existing node
+     * else: use the new node
+     *
+     * @param existingLocations Existing locations routed for the same table
+     */
+    public void updateLocations(Map<String, Map<String, List<Integer>>> existingLocations) {
+        if (existingLocations.equals(locations)) {
+            return;
+        }
+        // All existing shards
+        Map<Tuple<String, Integer>, String> existingShards = new HashMap<>();
```

Comment:
The alternative is https://gist.github.com/matriv/6a94e6f93fd3deec6d8ebd2019f9e3d4
which anyway requires another temp structure since we cannot add entries to a map while iterating over it.